### DISCUSSION TITLE
openEMS code structure

### DISCUSSION THREAD

**Discussion by LubomirJagos** [thliebig/openEMS-Project#123] [General]
Hello,
this is not any issue or similar thing, just place where I want to show you if anybody is interested how openEMS solver core is made and get information about tools how to inspect code to make it easier understand what is happening under hood. Also to show that solver itself is not black magic (alogrithm itself is black magic), I am myself not expert and don't know how to implement it (but my professors from university at this sentence will give me strange look :D ). So to the point.
OpenEMS solver core itself is written in C++ using object code approach (what is great) and there are tools for code inspection which makes life sooo easy to inspect code and grasp idea how is structured, one (and probably only one, if there is some else similar tool please please somebody write here it's name and url link) is open-source, unluckily nowadays not under development but it's running, it will be running and for inspection source codes like C++, JAVA, python more than capable, it's called Sourcetrail and it can be found here:
    https://github.com/CoatiSoftware/Sourcetrail/releases
    
Long story short, main thing how openEMS is solving EM problem is happening in these classes:
    1. openEMS::RunFDTD
    2. Engine::IterateTS
    3. Engine::UpdateVoltages and Engine::UpdateCurrents
    
That's all, it doesn't hurt :D Here below are screenshots from Sourcetrail how it looks like, I hope this topic give you at least new way how code can be inspected instead of boring reading texts :D

![image](https://github.com/thliebig/openEMS-Project/assets/18229825/df3cb83b-b965-4ddb-823b-2e01710d7e4e)
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/98df06c0-88b1-4e98-ac9c-253585c8d875)
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/b5929d86-deec-4a86-9ba7-47d5ebdf69e5)
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/c32dd0ea-ba78-4d6d-b413-d91e6bc45eb9)

**Comment by biergaizi** [comment]
>  Also to show that solver itself is not black magic (alogrithm itself is black magic)

I disagree, the algorithm is not black magic. All the algorithm does is doing some addition, subtraction, and multiplication per cell. This is finite difference, being "dumb" is its most appealing part... On the other hand, the math and physics derivations behind the update equations, including how the E and B field grids are interleaved, and how the R, L, G, C parameters and operators are calculated, are black magic ;-)

---

I have studied the openEMS engine in depth during my effort to develop CPU optimization and GPU acceleration. In short, it's a not difficult architecture, but some parts are quite tricky to understand if you're unaware of the overall architecture. Some typical questions are:

1. The relationship between the Engine and Operator implementation.

2. How an engine implementation is selected and launched.

3. How many engines are there, and what are the difference between each engine.

4. In the SSE version, how the data transpose is done.

6. How multi-threading and domain decomposition are performed.

7. How the extensions are invoked.

8. How the post-processing is invoked.

I have answers to all of these questions by now.

I really should send a pull request to document everything to help newcomers. But for now, just ask me personally if anyone's confused...

**Reply by 0xCoto** [reply]
About a year ago, we had initiated an effort to accelerate things on the GPU with CUDA, but our efforts fell short by the time we had to deal with these exact questions. You've definitely explored the codebase quite deeper than us by now though, so you'll hopefully be able to figure something out. :)

I might start investing some time into it in the coming months if nothing else (openEMS-related) comes up.

**Reply by LubomirJagos** [reply]
thanks for reaction, I put there black magic to get you hooked into this conversion, both of you who I wanted to be hooked here are already here, so you are getting notifications :)

@biergaizi already wrote there main topics which your brain already chews and you have your mental map of code, that would be nice since you already know how it's running to have some UML diagram and some sequence diagrams ie. using www.draw.io or another tool

I started this discussion since openEMS code structure is as you wrote not so complicated, but to get orientation in it require some code browsing, since now I wasn't really deeply interested how it's implemented, just once or two times opened it and looked what is there and then I become more interested as you posted your inductance extension which till now I haven't tested

If there would be some documentation how things are done that would be nice, I appreciate your approach that you are open to questions but most of time even myself requires lot of time to even create right question which is valid, now I can inspect code using sourcetrail and look what it's doing, it would be nice if there would be at least some simple overview for one thread workflow over some simple planar antenna which will show what is happening under hood in code.

I know that people feel that everything must be perfect, it doesn't need to be perfect, even whole .txt document where are notes is great, when I'm working on my GUI fronted for openEMS it's always iterative work, write something and then more and more and it's getting shape, so I think if you can provide .txt file with everything what you want it's good start point :)

**Reply by LubomirJagos** [reply]
I love memes.
![0cc99dc0c8da7fc5cefc9a3e1f75257b--big-thing-michael-fassbender](https://github.com/thliebig/openEMS-Project/assets/18229825/824ea4ef-f498-4929-b2a6-c67742578995)

**Reply by gadiLhv** [reply]
The opration of the library is pretty straight forward, as you suggested, @LubomirJagos.

1. openEMS::RunFDTD
2. Engine::IterateTS
3. Engine::UpdateVoltages and Engine::UpdateCurrents

Are part of it. In `IterateTS` there are pre-process and post-process calls, that have to do with all the engine extensions. 
Most of the trouble I had (and still having) along the way are actually during setup. In each operator extension there is always a method called `BuildExtension`, which actually creates all the variables and coefficients necessary for the engine extension runtime.

In the functions I've worked on, I've put a lot of energy on figuring out what's actually going on in these specific methods.

Proper documentation will take a lot of time and effort for two reason:
1. It's a buttload of code.
2. Proper mathematical referencing is a must.

Not saying it's impossible, but it is deterring.
You are welcome to look at [this pull request](https://github.com/thliebig/openEMS/pull/121) . Imagine formulating and attaching one of these to each engine extension.

However, as far as black magic goes... FDTD (and in this EC-FDTD is included) is the most straight-forward, simple to understand and implement method. Variational methods and weak formulations (Finite element stuff)... I've studied these for years. Can't say I fully understand them to this day. Spurious solutions, Ugh...

Feel free to ask questions. I've been out and about on the library for a short while.
